### PR TITLE
Demonstrate some existing issues

### PR DESCRIPTION
**Pull request by marksto**
@weavejester Hello James!

I've created this fake PR to showcase existing issues and other quirks lurking in the current `resauce` implementation.

I'm using this lib's features for quite a while now. At some point I started to came across different issues and corner cases, so I had to switch to using a local patched version of `resauce`. But I think it's a good time to contribute these fixes back.

I as well decided to save you and me some time here and provide the findings all at once, as a single demonstrative PR (and not a series of open issues), which we can discuss in one place. We can open dedicated issues later, if needed. I hope you don't mind.

---

To demonstrate the findings I first grab the modern `fs` lib, update the Clojure version (to the min required for `fs` to work), and introduce a substantial test coverage (that extends the one that was already there). In these new unit tests I do my best to explore all possible corner cases and resource representations that we have in Java/Clojure nowadays (incl. `Path`s).

The test results are added in a separate file. I'll go through them one by one in the comments below.

Cheers,
Mark

### CODE REVIEW

File: test-results.log [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      false
+      true
+      false
+      false
+      true
+      true
+      false
+      true
+      false
+      true)))
```
Comment (marksto) [inline test-results.log:30]: This happens for `correct-fs-dirs`.

This is kinda major. Perfectly valid inputs sometimes get misclassified by the mm `url-scheme` dispatcher fn due to an incorrect/incomplete implementation.

```
(correct-fs-dirs)
=>
(#object [java.io.File 0x5163fef3 "/Users/me"] ; <~ false
 #object [sun.nio.fs.UnixPath 0x1f2bf9ef "/Users/me"] ; <~ false
 #object [java.net.URL 0x66c9be6a "file:/Users/me/"]
 #object [java.net.URI 0x6beb06f2 "file:///Users/me/"]
 "/Users/me" ; <~ false
 "file:/Users/me"
 "/Users/me/" ; <~ false
 "file:/Users/me/"
 #object [java.io.File 0x5b8a6621 "/Users/me/Desktop/Projects/OSS/resauce/test/resauce"] ; <~ false
 #object [sun.nio.fs.UnixPath 0x95233fe "/Users/me/Desktop/Projects/OSS/resauce/test/resauce"] ; <~ false
 #object [java.net.URL 0x5574e811 "file:/Users/me/Desktop/Projects/OSS/resauce/test/resauce/"]
 #object [java.net.URI 0x4c41e8e3 "file:///Users/me/Desktop/Projects/OSS/resauce/test/resauce/"]
 "/Users/me/Desktop/Projects/OSS/resauce/test/resauce" ; <~ false
 "file:/Users/me/Desktop/Projects/OSS/resauce/test/resauce"
 "/Users/me/Desktop/Projects/OSS/resauce/test/resauce/" ; <~ false
 "file:/Users/me/Desktop/Projects/OSS/resauce/test/resauce/")
```

File: test-results.log [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+java.lang.IllegalArgumentException: No implementation of method: :as-url of protocol: #'clojure.java.io/Coercions found for class: sun.nio.fs.UnixPath
+at clojure.core$_cache_protocol_fn.invokeStatic (core_deftype.clj:583)
+clojure.java.io$fn__10827$G__10822__10832.invoke (io.clj:35)
+resauce.core$url_file.invokeStatic (core.clj:25)
+resauce.core$url_file.invoke (core.clj:24)
+resauce.core$eval607$fn__608.invoke (core.clj:33)
+...
+resauce.core_test$fn__669$fn__678.invoke (core_test.clj:165)
+resauce.core_test$fn__669.invokeStatic (core_test.clj:163)
+resauce.core_test/fn (core_test.clj:158)
```
Comment (marksto) [inline test-results.log:72]: This happens for `incorrect-fs-dirs`.

While one can argue about whether it should return a `false` result or throw here, there is no doubt that it would be great to have full support for Java NIO `Path`.

Arguments for returning `false` rather than throwing:
- current impl does not throw for obviously wrong inputs, e.g. `(directory? nil)`
- a non-throwing version has a better DX, e.g. it can be used in threading as is

File: test-results.log [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+Assertion failed with error:
+resources inside a JAR file
+
+java.lang.NullPointerException: Cannot invoke "String.endsWith(String)" because "s" is null
+at resauce.core$add_ending_slash.invokeStatic (core.clj:9)
+resauce.core$add_ending_slash.invoke (core.clj:8)
+resauce.core$eval612$fn__613.invoke (core.clj:40)
+...
+resauce.core_test$fn__669.invokeStatic (core_test.clj:171)
+resauce.core_test/fn (core_test.clj:158)
```
Comment (marksto) [inline test-results.log:121]: This happens for `incorrect-jar-dirs`.

While one can argue about whether it should return a `false` result or throw here, there is no doubt that it should not end up with an NPE.

Arguments for returning `false` rather than throwing:
- current impl does not throw for obviously wrong inputs, e.g. `(directory? nil)`
- a non-throwing version has a better DX, e.g. it can be used in threading as is

File: test-results.log [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+Actual:
+(not
+  (every?
+    false?
+    (nil
+      nil
+      nil
+      nil
+      nil
+      nil)))
```
Comment (marksto) [inline test-results.log:103]: This happens for `files-in-jars`.

This is kinda picky, but it would be nice to return `false` for files in JARs, just like for any regular FS files.

File: test/resauce/core_test.clj [hunk 5]
```diff
@@ -1,14 +1,216 @@
...
+   f))
+
+(defn some-file-path ^Path []
+  (->> (fs/list-dir (fs/home))
+       (filter fs/regular-file?)
+       (first)))
+
+(defn some-file-url ^URL []
+  ;; NB: Use some concrete resource that we know lives on the classpath as file.
+  (io/resource "resauce/core.clj"))
```
Comment (marksto) [inline test/resauce/core_test.clj:86]: An already used `"resauce/core.clj"` file is complemented by any file from the home directory.

File: test/resauce/core_test.clj [hunk 6]
```diff
@@ -1,14 +1,216 @@
...
+          [(some-file-path)
+           (some-file-url)]))
+
+(defn- ->dir-endings [f]
+  ((juxt identity #(str % "/")) (str f)))
+
+(defn some-fs-dirs []
+  ;; NB: Use some concrete resources that we know live on the classpath as dirs.
+  [(str (fs/home))
+   (str (io/as-file (io/resource "resauce")))])
```
Comment (marksto) [inline test/resauce/core_test.clj:99]: An already used `"resauce"` dir is complemented by the home directory.

File: test/resauce/core_test.clj [hunk 7]
```diff
@@ -1,14 +1,216 @@
...
+
+(defn- ->jar-meta-inf-dir
+  "Yields 'jar:file:/<...>.jar!/META-INF' string."
+  [^URL jar-url]
+  (str "jar:" (->jar-file jar-url) "!/META-INF"))
+
+(defn some-jar-file-urls []
+  ;; NB: Use some concrete resources that we know live on the classpath in JARs.
+  [(io/resource "clojure/core.clj")
+   (io/resource "META-INF/leiningen/medley/medley/README.md")])
```
Comment (marksto) [inline test/resauce/core_test.clj:144]: An already used `"clojure/core.clj"` JAR file is complemented by the `medley`'s README, also JARed.

File: test/resauce/core_test.clj [hunk 8]
```diff
@@ -1,14 +1,216 @@
...
+     (string? f)
+     (juxt io/as-url
+           str->uri
+           identity))
+   f))
+
+(defn some-jar-dir-urls []
+  ;; NB: Use some concrete resources that we know live on the classpath in JARs.
+  [(io/resource "clojure")
+   (io/resource "medley")])
```
Comment (marksto) [inline test/resauce/core_test.clj:177]: An already used `"clojure"` JAR dir is complemented by the `"medley"` dir, also JARed.

**Review by marksto — COMMENTED** [review]
So, here are the main things to pay attention to. Everything else here is test utilities that deal with transforming "resource-namish" things (URLs, URIs, Files, Paths, and Strings) in every possible way prior to feeding them to the function under test.

**Comment by weavejester** [comment]
Thanks for the 'PR' and all your work on this. However, this PR is pretty difficult for me to parse. Could you open individual issues or PRs with examples of what resauce produces and what it *should* produce?

Please bear in mind I haven't touched this project in four years.

**Comment by marksto** [comment]
> Could you open individual issues or PRs with examples of what resauce produces and what it should produce?

Nah. I'd rather live with and maintain a patched version myself.
I understand how much is on your plate, James, so no worries.

**Comment by weavejester** [comment]
Maintaining a fork is even better, as I don't have to do any work ;)
